Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,270 @@
+}
+
+extension CachedFramework: Decodable {
+	static func decode(_ j: JSON) -> Decoded<CachedFramework> {
+		return curry(self.init)
+			<^> j <| CachedFramework.nameKey
+			<*> j <| CachedFramework.md5Key
+	}
+}
+
+struct VersionFile {
+	let commitish: String
+	let xcodeVersion: String
+	
+	let macOS: [CachedFramework]?
+	let iOS: [CachedFramework]?
+	let watchOS: [CachedFramework]?
+	let tvOS: [CachedFramework]?
+	
+	static let commitishKey = "commitish"
+	static let xcodeVersionKey = "xcodeVersion"
+
+	subscript(_ platform: Platform) -> [CachedFramework]? {
+		switch platform {
+		case .macOS:
+			return macOS
+		case .iOS:
+			return iOS
+		case .watchOS:
+			return watchOS
+		case .tvOS:
+			return tvOS
+		}
+	}
+	
+	func toJSONObject() -> Any {
+		var dict: [String: Any] = [
+			VersionFile.commitishKey : commitish,
+			VersionFile.xcodeVersionKey : xcodeVersion
+		]
+		for platform in Platform.supportedPlatforms {
+			if let caches = self[platform] {
+				dict[platform.rawValue] = caches.map { $0.toJSONObject() }
+			}
+		}
+		return dict
+	}
+	
+	init(commitish: String, xcodeVersion: String, macOS: [CachedFramework]?, iOS: [CachedFramework]?, watchOS: [CachedFramework]?, tvOS: [CachedFramework]?) {
+		self.commitish = commitish
+		self.xcodeVersion = xcodeVersion
+		
+		self.macOS = macOS
+		self.iOS = iOS
+		self.watchOS = watchOS
+		self.tvOS = tvOS
+	}
+	
+	init?(url: URL) {
+		guard FileManager.default.fileExists(atPath: url.path),
+			let jsonData = try? Data(contentsOf: url),
+			let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments),
+			let versionFile: VersionFile = Argo.decode(json) else {
+			return nil
+		}
+		self = versionFile
+	}
+	
+	func md5s(for cachedFrameworks: [CachedFramework], platform: Platform, binariesDirectoryURL: URL) -> SignalProducer<String?, CarthageError> {
+		return SignalProducer(cachedFrameworks)
+			.flatMap(.concat) { cachedFramework -> SignalProducer<String?, CarthageError> in
+				let frameworkName = cachedFramework.name
+				let frameworkBinaryURL = binariesDirectoryURL
+					.appendingPathComponent(platform.rawValue, isDirectory: true)
+					.resolvingSymlinksInPath()
+					.appendingPathComponent("\(frameworkName).framework", isDirectory: true)
+					.appendingPathComponent("\(frameworkName)", isDirectory: false)
+				return md5ForFileAtURL(frameworkBinaryURL)
+					.map { md5 -> String? in
+						return md5
+					}
+					.flatMapError { _ in
+						return SignalProducer(value: nil)
+					}
+			}
+	}
+
+	func satisfies(platform: Platform, commitish: String, xcodeVersion: String, binariesDirectoryURL: URL) -> SignalProducer<Bool, CarthageError> {
+		guard let cachedFrameworks = self[platform] else {
+			return SignalProducer(value: false)
+		}
+		
+		let md5s = self.md5s(for: cachedFrameworks, platform: platform, binariesDirectoryURL: binariesDirectoryURL)
+		return self.satisfies(platform: platform, commitish: commitish, xcodeVersion: xcodeVersion, md5s: md5s)
+	}
```

Comment:
Let's make the `satisfies` right below this take a `md5s: [String?]` instead of a `SignalProducer`.

```swift
return self
    .md5s(for: cachedFrameworks, platform: platform, binariesDirectoryURL: binariesDirectoryURL)
    .collect()
    .map { md5s in
        return self.satisfies(platform: platform, commitish: commitish, xcodeVersion: xcodeVersion: md5s: md5s)
    }

func satisfies(platform: Platform, commitish: String, xcodeVersion: String, md5s: [String?]) -> Bool {
    guard let cachedFrameworks = self[platform], commitish == self.commitish, xcodeVersion == self.xcodeVersion else {
        return SignalProducer(value: false)
    }
    
     return md5s == cachedFrameworks.map { $0.md5 }
}
```

---

Although this really makes we think we should do this instead:

1. Add a function that creates a `VersionFile` from the state on disk
2. Add a method that asks whether one `VersionFile` is satisfied by another
3. Here, load the `VersionFile` from disk (1) and ask whether it satisfies (2)
4. Later, use (1) to persist the new version file

Essentially, `createVersionFile` below would return a `SignalProducer<VersionFile, CarthageError>` and _not_ write out the file to disk. It could then be reused for both writing the version file and checking whether it's up to date.